lib/ansible/modules/remote_management/manageiq/manageiq_dynamic_resource_instance.py

```diff
@@ -0,0 +1,381 @@
+
+        try:
+            result = self.client.post(url, action='delete')
+        except Exception as e:
+            self.module.fail_json(
+                msg="failed to delete dynamic resource instance %s: %s" % (
+                    dynamic_resource_instance.get('name'), str(e)
+                )
+            )
+
+        return dict(
+            changed=True,
+            msg="successfully deleted the dynamic resource instance %s" % (
+                dynamic_resource_instance.get('name')
+            )
+        )
+
+
+def main():
+    argument_spec = dict(
+        name=dict(required=True, type='str'),
+        state=dict(choices=['absent', 'present'], default='present'),
+        dynamic_resource_definition=dict(type='str'),
+        property_attributes=dict(type='dict'),
+        services=dict(type='list', default=[]),
+        providers=dict(type='list', default=[]),
+    )
+
+    required_if = [
+        (
+            'state', 'present', (
+                'dynamic_resource_definition', 'property_attributes',
+            )
+        )
+    ]
+    # add the manageiq connection arguments to the arguments
+    argument_spec.update(manageiq_argument_spec())
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        required_if=required_if,
+    )
+
+    name = module.params['name']
+    state = module.params['state']
+    dynamic_resource_definition = module.params['dynamic_resource_definition']
+    property_attributes = module.params['property_attributes']
+    services = module.params['services']
+    providers = module.params['providers']
+
+    manageiq = ManageIQ(module)
+    manageiq_dynamic_resource_instance = ManageIQDynamicResourceInstance(manageiq)
+
+    # Check if definition exists
+    dynamic_resource_definition = manageiq_dynamic_resource_instance.dynamic_resource_definition(
+        dynamic_resource_definition
+    )
+    if not dynamic_resource_definition and state != "absent":
+        module.fail_json(
+            msg="Dymamic resource definition:%s does not exist." % (dynamic_resource_definition)
+        )
+    # Get instance if exists
+    dynamic_resource_instance = manageiq_dynamic_resource_instance.dynamic_resource_instance(
+        name
+    )
+
+    if state == "absent":
+        # if we have a instance, delete it
+        if dynamic_resource_instance:
+            res_args = manageiq_dynamic_resource_instance.delete_instance(
+                dynamic_resource_instance
+            )
+        # if we do not have a instance, nothing to do
+        else:
+            res_args = dict(
+                changed=False,
+                msg="dynamic resource instance  %s: does not exist in manageiq" % (name)
+            )
+
+    if state == "present":
+        # Validate that providers and services exist
+        service_objs = [
+            manageiq_dynamic_resource_instance.service(service) for service in services
```

Comment:
This call is a good example how our existing pattern of stuffing everything into one class named after the thing the module manipulates is getting awkward...
- If you agree and see a better name than `manageiq_dynamic_resource_instance`, feel free to rename it (and/or the class).
- In this case you don't need any state from the class, only the `ManageIQ` API wrapper.
  Consider inlining here `manageiq.find_collection_resource_by('services', name=name)` and getting rid of the method.

(same for `provider` method call below)

Reply:
I was just keeping the naming consistent with all the other modules.
To make it a bit cleaner I will just use the api object to fetch it, instead of wrapping it in a method.

Reply:
It is also OK as-is if you like it; I just wanted to make it clear you don't *have* to obey the consistency if you see a better way.

[I think I was in a bike-shedding mood when I did this review...  Feel free to push back :-)]